### PR TITLE
Use IPLD-prime: target merge branch

### PR DESCRIPTION
**Pull request by hannahhoward**
superseeds #64 

- switch to go-codec-dagpb
- switch to go-ipld-prime linksystem branch

### CODE REVIEW

File: node.go [hunk 1]
```diff
@@ -382,3 +390,17 @@ func (n *ProtoNode) Tree(p string, depth int) []string {
 	}
 	return out
 }
+
+func ProtoNodeConverter(b blocks.Block, nd ipld.Node) (legacy.UniversalNode, error) {
```
[the remark this answers is not in the document]
Comment (hannahhoward) [reply node.go:394]: The main use of this code is actually here: https://github.com/ipfs/go-merkledag/pull/67/files#diff-5e7919ad01b60e6e687fb533052ee768a19c08aa341016c5d2066f539c89bf52R28

Which feeds into here: https://github.com/ipfs/go-ipld-legacy/blob/master/decode.go#L44

The basic problem again is we need a node and block to satisfy the old interface. In the case of interacting with the block service in the way we do in go-merkledag, we end up with a block first. We then assemble a prime node from the block, the wrap both in a legacy node structure of some kind. The reason for the registry is that the ProtoNode (a legacy node implementation) in this repository has to have some special abilities -- namely that you can add and remove links and reserialize it. 

BTW, upon further reflection I believe I can remove the usage of this code from the go-path PR in go-ipfs.

The reason for having a cached copy of the the marshalled data in DagPB is that this struct is mutable, so you might mutate and remarshal. In ipld-prime, a node is never mutable -- it's a seperate interface from a nodebuilder, which is how you can actually construct and mutate nodes. I don't see why you would ever keep marshalled data around -- if you've marshalled it once, you're not going to marshall it again until you construct a new node. You may want to talk to @warpfork about this one.

[the remark this answers is not in the document]
Comment (hannahhoward) [reply node.go:394]: upon further further reflection it's not clear whether it's possible to remove this code from the go-ipfs PR :(, at least till we juse move everything over to the new stuff.

**Comment by warpfork** [comment]
Probably known, but noting out loud: this needs a rebase onto master (or some manual merging, owner's choice).

However, it doesn't look like it should be too nasty.

- most of the changes are to CI infra, which hopefully won't conflict in interesting ways.
- https://github.com/ipfs/go-merkledag/pull/72 might be a little scary (it downgrades a dep?)... but it looks like the deps it changed are not changed for load-bearing reasons in this PR (afaict), so while it creates a conflict in `go.mod`, hopefully it's just a manual resolve in that file only.

File: node.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (warpfork) [reply node.go:394]: Sounds like this is mostly to satisfy legacy interfaces, aka control the scope for this multi-repo merge party we've got coming up, and so it's here to stay, _for now_.

We can refresh ourselves on what the best path forward will be when we get another large block of work time to start chopping out legacy stuff.  (Sounds like we might end up in a position where we later try to remove several swaths at once, and it might end up easier that way.  But who knows.)  For now, I'd say "let's document the intent and move on", but even that seems pretty strongly hinted already by the func signature returning something called literally `legacy`!

So: I think we can call this resolved for now?  I'll mark it as such.

File: merkledag.go [hunk 2]
```diff
@@ -10,16 +10,23 @@ import (
...
+	format "github.com/ipfs/go-ipld-format"
+	legacy "github.com/ipfs/go-ipld-legacy"
+	dagpb "github.com/ipld/go-codec-dagpb"
+	_ "github.com/ipld/go-ipld-prime/codec/raw"
+	basicnode "github.com/ipld/go-ipld-prime/node/basic"
 )
 
 // TODO: We should move these registrations elsewhere. Really, most of the IPLD
 // functionality should go in a `go-ipld` repo but that will take a lot of work
 // and design.
```
Comment (warpfork) [inline merkledag.go:22]: I'd just like to reflect for a moment, with gentle amusement, that the blame on this is about 4 years old.

It took awhile, but we're getting there now :)

(We'll just probably see this whole library repo disappear at the same time as this particular block of registration calls disappears and the TODO becomes finished, heh.)

File: node.go [hunk 3]
```diff
@@ -16,16 +20,20 @@ var (
...
-	links []*ipld.Link
+	links []*format.Link
 	data  []byte
 
 	// cache encoded/marshaled value
-	encoded []byte
-
-	cached cid.Cid
+	encoded *immutableProtoNode
+	cached  cid.Cid
```
Comment (warpfork) [inline node.go:36]: I think I need a comment block on these types that briefly describes what the difference is between these and [go-codec-dagpb.PBNode](https://pkg.go.dev/github.com/ipld/go-codec-dagpb#PBNode).

Maybe it's all legacy, maybe it's actually somewhat functional...?  It seems like it will be hard for a reader to tell without a large amount of splunking.

File: prime.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+import (
+	"github.com/ipld/go-ipld-prime"
+	dagpb "github.com/ipld/go-codec-dagpb"
+)
+
+// Kind returns a value from the Kind enum describing what the
+// essential serializable kind of this node is (map, list, integer, etc).
+// Most other handling of a node requires first switching upon the kind.
+func (n *ProtoNode) Kind() ipld.Kind {
```
Comment (warpfork) [inline prime.go:11]: The volume of boilerplate in this file is the main reason I wonder what's the diff between this type and the `go-codec-dagpb.PBNode` type [in my other comment](https://github.com/ipfs/go-merkledag/pull/67/files#r682629402).

It doesn't seem especially harmful to have this if it's providing a useful service, but I don't immediately know what why/what that service is.

**Comment by warpfork** [comment]
I _think_ I'm mostly okay with this?

Per reasoning in [the big review of repos and goalposts for this mergeparty I made over in `go-path`](https://github.com/ipfs/go-path/pull/36#issuecomment-892636347), the aim here is moving forward, and getting new code to do more load-bearing... and while focusing on that, we are conserving scope by leaving stuff in this repo as facades, and are specifically _not_ polishing these APIs in any way.  So the result is not going to be pretty, yet.  And that's okay.  We're just trying to be in a better place for the next work cycle.

AFAICT, checklist to get this ready for mergeparty:

- [ ] Rebase onto master (or merge master in and resolve).
- [ ] Some kind of doc comment on ProtoNode type to disambiguate vs the other similarly-named types in other repos, please
- [ ] Should we... update the README to talk about alternatives folks should look to?  Or does that remain unclear?  (I just now noticed how much of it is a TODO section already, and it does even already say "deprecate this package"(!) so we can... keep with tradition, heh.)

File: .travis.yml [hunk 5]
```diff
@@ -4,7 +4,7 @@ os:
 language: go
 
 go:
-  - 1.11.x
+  - 1.14.x
```
Comment (aschmahmann) [inline .travis.yml:7]: @hannahhoward it seems like we need to be rebased on master since we're not using travis anymore.

File: coding_test.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+			Cid:  someCid,
+		})
+	}
+
+	enc, err := node.EncodeProtobuf(true)
+	if err != nil {
+		panic(err)
+	}
+	benchInput = enc
+	// println(len(benchInput))
```
Comment (aschmahmann) [inline coding_test.go:33]: remove?

File: coding_test.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+	cid "github.com/ipfs/go-cid"
+	ipld "github.com/ipfs/go-ipld-format"
+	"github.com/ipfs/go-merkledag"
+)
+
+var benchInput []byte
+
+func init() {
+	someData := bytes.Repeat([]byte("some plaintext data\n"), 10)
+	someCid, _ := cid.Cast([]byte{1, 85, 0, 5, 0, 1, 2, 3, 4})
```
Comment (aschmahmann) [inline coding_test.go:17]: what is this and why are we skipping the error check?

File: coding_test.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+			if err != nil {
+				b.Fatal(err)
+			}
+
+			enc, err := node.EncodeProtobuf(true)
+			if err != nil {
+				b.Fatal(err)
+			}
+			// println(len(benchInput), len(enc))
+			_ = enc
```
Comment (aschmahmann) [inline coding_test.go:50]: remove?

File: go.mod [hunk 9]
```diff
@@ -2,16 +2,23 @@ module github.com/ipfs/go-merkledag
 
 require (
 	github.com/gogo/protobuf v1.3.1
-	github.com/ipfs/go-block-format v0.0.2
-	github.com/ipfs/go-blockservice v0.1.0
-	github.com/ipfs/go-cid v0.0.3
-	github.com/ipfs/go-datastore v0.3.1
-	github.com/ipfs/go-ipfs-blockstore v0.1.0
+	github.com/gopherjs/gopherjs v0.0.0-20190430165422-3e4dfb77656c // indirect
```
Comment (aschmahmann) [inline go.mod:5]: we've got some indirect dependencies we probably want to remove

File: merkledag.go [hunk 10]
```diff
@@ -10,16 +10,23 @@ import (
 	bserv "github.com/ipfs/go-blockservice"
 	cid "github.com/ipfs/go-cid"
 	ipldcbor "github.com/ipfs/go-ipld-cbor"
-	ipld "github.com/ipfs/go-ipld-format"
+	format "github.com/ipfs/go-ipld-format"
+	legacy "github.com/ipfs/go-ipld-legacy"
+	dagpb "github.com/ipld/go-codec-dagpb"
+	_ "github.com/ipld/go-ipld-prime/codec/raw"
```
Comment (aschmahmann) [inline merkledag.go:16]: indirect imports should generally come with comments explaining why they're there so the next person knows what's up 😄.

File: merkledag.go [hunk 11]
```diff
@@ -10,16 +10,23 @@ import (
...
 // TODO: We should move these registrations elsewhere. Really, most of the IPLD
 // functionality should go in a `go-ipld` repo but that will take a lot of work
 // and design.
 func init() {
-	ipld.Register(cid.DagProtobuf, DecodeProtobufBlock)
-	ipld.Register(cid.Raw, DecodeRawBlock)
-	ipld.Register(cid.DagCBOR, ipldcbor.DecodeBlock)
+	format.Register(cid.DagProtobuf, DecodeProtobufBlock)
+	format.Register(cid.Raw, DecodeRawBlock)
+	format.Register(cid.DagCBOR, ipldcbor.DecodeBlock)
```
Comment (aschmahmann) [inline merkledag.go:26]: Why was this here before and do we need it now? Mostly noticing the asymmetry in that the other two codecs have legacy types and this one doesn't.

File: node.go [hunk 12]
```diff
@@ -204,9 +210,11 @@ func (n *ProtoNode) Copy() ipld.Node {
 	return nnode
 }
 
-// RawData returns the protobuf-encoded version of the node.
 func (n *ProtoNode) RawData() []byte {
-	out, _ := n.EncodeProtobuf(false)
+	out, err := n.EncodeProtobuf(false)
+	if err != nil {
+		panic(err)
+	}
```
Comment (aschmahmann) [inline node.go:217]: It looks like we previously just ignored the errors. While I'd generally rather panic than silently ignore errors I'd like to know what can cause an error here and if any of it is user/network (rather than developer) issues.

File: node_test.go [hunk 13]
```diff
@@ -12,15 +12,25 @@ import (
 	ipld "github.com/ipfs/go-ipld-format"
 )
 
+var sampleCid cid.Cid
+
+func init() {
+	var err error
+	sampleCid, err = cid.Cast([]byte{1, 85, 0, 5, 0, 1, 2, 3, 4})
```
Comment (aschmahmann) [inline node_test.go:19]: what is this?

File: node_test.go [hunk 14]
```diff
@@ -12,15 +12,25 @@ import (
...
 	nd.SetLinks([]*ipld.Link{
-		{Name: "a"},
-		{Name: "b"},
-		{Name: "c"},
+		{Name: "a", Cid: sampleCid},
+		{Name: "b", Cid: sampleCid},
+		{Name: "c", Cid: sampleCid},
 	})
-	expected, err := cid.Decode("QmSN3WED2xPLbYvBbfvew2ZLtui8EbFYYcbfkpKH5jwG9C")
+	expected, err := cid.Decode("QmciCHWD9Q47VPX6naY3XsPZGnqVqbedAniGCcaHjBaCri")
```
Comment (aschmahmann) [inline node_test.go:33]: was the prior test valid DagPB? If so, are we get the same CID output as a way of making sure we haven't broken things?

What's the idea behind the change here? Is it just that we're testing more here? Same for the other modifications in this file.

File: raw.go [hunk 15]
```diff
@@ -100,4 +116,11 @@ func (rn *RawNode) MarshalJSON() ([]byte, error) {
 	return json.Marshal(string(rn.RawData()))
 }
 
-var _ ipld.Node = (*RawNode)(nil)
+func RawNodeConverter(b blocks.Block, nd ipld.Node) (legacy.UniversalNode, error) {
+	if nd.Kind() != ipld.Kind_Bytes {
+		return nil, ErrNotProtobuf
```
Comment (aschmahmann) [inline raw.go:121]: This seems like the wrong error

**Review by aschmahmann — CHANGES_REQUESTED** [review]
Not many blocking issues, but this code is fairly tricky so I've asked for clarification in a number of places. Also, we need to rebase on master since we're out of date

File: coding.go [hunk 16]
```diff
@@ -22,38 +25,84 @@ const _ = pb.DoNotUpgradeFileEverItWillChangeYourHashes
...
-	for i, l := range pbnl {
-		n.links[i] = &ipld.Link{Name: l.GetName(), Size: l.GetTsize()}
-		c, err := cid.Cast(l.GetHash())
-		if err != nil {
-			return fmt.Errorf("link hash #%d is not valid multihash. %v", i, err)
+func fromImmutableNode(encoded *immutableProtoNode) *ProtoNode {
+	n := new(ProtoNode)
+	n.encoded = encoded
+	if n.encoded.PBNode.Data.Exists() {
+		n.data = n.encoded.PBNode.Data.Must().Bytes()
```
[the remark this answers is not in the document]
Comment (hannahhoward) [reply coding.go:41]: the if check immediately preceding gaurauntees no panic -- if it Exists returns true, then Must always succeeds.

File: coding.go [hunk 17]
```diff
@@ -22,38 +25,84 @@ const _ = pb.DoNotUpgradeFileEverItWillChangeYourHashes
...
+	linkAllocs := make([]format.Link, numLinks)
+	for i := int64(0); i < numLinks; i++ {
+		next := n.encoded.PBNode.Links.Lookup(i)
+		name := ""
+		if next.FieldName().Exists() {
+			name = next.FieldName().Must().String()
 		}
-		n.links[i].Cid = c
+		c := cid.Undef
+		c = next.FieldHash().Link().(cidlink.Link).Cid
```
[the remark this answers is not in the document]
Comment (hannahhoward) [reply coding.go:53]: honestly, I consider "second link interface" far enough off not to care that much. Per @warpfork for the time being it's fine to just do casts like this.

File: coding_test.go (hunk 7 above)
Comment (hannahhoward) [reply coding_test.go:17]: as I understand it, all we're trying to do here is make a test protonode with 10 links, and we don't really care about the content. @mvdan wrote this code though

File: merkledag.go (hunk 11 above)
Comment (hannahhoward) [reply merkledag.go:26]: I honestly have no idea why DagCBOR is being registered in a library that contains only a protobuf implementation. But I don't know the history.

But these are needed for go-ipld-format to continue working I think.

As for the legacy types, there is a default legacy type implementation in go-ipld-legacy that will work for CBOR and other formats.

The reason these special types were needed is that they implement MORE than the go-ipld-format spec and the logic of some of the go-ipld-format methods differs from the standard implementation (namely Resolve & ResolveLink which search the links for a matching name)

File: node.go (hunk 12 above)
Comment (hannahhoward) [reply node.go:217]: I'm going with "that's what it does everywhere else in the code and this brings it in line"? But you'll have to ask @mvdan cause he's on the blame. But I don't feel attached.

File: node_test.go (hunk 14 above)
Comment (hannahhoward) [reply node_test.go:33]: My understanding that go-codec-dagpb does not accept Links with empty CIDS. This is inline with the DagPB spec, which states:
"Hash:
- even though Hash is optional in the Protobuf encoding, it should not be treated as optional when creating new blocks or decoding existing ones, an omitted Hash should be interpreted as a bad block"

I suspect the old library accepted empty CIDs because it was written earlier. I agree though this may be of concern -- at least for compatibility checking, even if the old implementation is TOO flexible for the spec.

**Comment by hannahhoward** [comment]
@aschmahmann I believe I've addressed most comments and this is ready for re-review

File: raw.go (hunk 15 above)
Comment (hannahhoward) [reply raw.go]: changed

File: coding.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (warpfork) [reply coding.go:53]: Yeah.  I consider this fine until we actually seriously expect existence of another possible inhabitant.  We don't, today; and if we ever add one, then we will just have to make a point to walk through all these repos and add real branches for this (and it should be easy to do by just grepping for casts).

(Big picture? I'm often regretting the `Link` interface at all.  The hope I had long ago was that that interface would help us discover what bits of the CID type _really matter_ vs what are historically accumulated noise, and also keep go-cid's transitive dep sprawl from going everywhere all the time.  Those goals still don't seem wrong, and may have even been partially successful... but it's certainly dubious whether that those discoveries were gained cost-effectively, and we still end up casting out to a concrete form all the time because the interface still doesn't describe all the functionality we regularly need.  Code in `go-ipld-prime` often manages to work using interface alone (success), but our applications tend not to (not success).  Now: moving in _any_ direction at all on this requires careful change planning, and is well _well_ beyond what we can think about bringing up into scope for any of these PRs.  But meanwhile these casts seem pretty tame to me.)

File: coding_test.go (hunk 7 above)
Comment (mvdan) [reply coding_test.go:17]: yep it's just a sample dummy CID for use in tests.

File: node.go (hunk 12 above)
Comment (mvdan) [reply node.go:230]: An error (and thus panic) can happen if the IPLD node is malformed, or if the checksum for the CID fails. Both should basically happen never, unless something went terribly wrong, like memory corruption or an internal bug.

**Review by Topyoutacrscrypto — APPROVED** [review]

File: coding_test.go (hunk 7 above)
Comment (aschmahmann) [reply coding_test.go:17]: seems fine, but some explanation would be great (i.e. I think this is a CIDv1 with a raw codec, and the identity multihash with length 5 and the value `[]{0, 1, 2, 3, 4}`).

File: node_test.go (hunk 14 above)
Comment (aschmahmann) [reply node_test.go:33]: Seems like a reasonable breakage to me, I'd certainly hope this couldn't matter in go-ipfs given that it's sort of non-sensical and spec-breaking.

File: node.go (hunk 12 above)
Comment (aschmahmann) [reply node.go:230]: ok, as long as it isn't I/O or user based this is probably fine.

File: node_test.go (hunk 13 above)
Comment (hannahhoward) [reply node_test.go:20]: address

File: merkledag.go (hunk 11 above)
Comment (aschmahmann) [reply merkledag.go:26]: Conclusion: We're leaving these in as convenience functions for people who have been relying on registration happening here.

TODO: Let's file an issue for removing these

**Review by aschmahmann — APPROVED** [review]

File: node.go [hunk 18]
```diff
@@ -278,8 +299,8 @@ func (n *ProtoNode) Loggable() map[string]interface{} {
 // UnmarshalJSON reads the node fields from a JSON-encoded byte slice.
 func (n *ProtoNode) UnmarshalJSON(b []byte) error {
 	s := struct {
-		Data  []byte       `json:"data"`
-		Links []*ipld.Link `json:"links"`
+		Data  []byte         `json:"data"`
+		Links []*format.Link `json:"links"`
 	}{}
```
Comment (aschmahmann) [inline node.go:304]: @warpfork @rvagg @willscott does this need to be changed to `Data` and `Links`? Perhaps relatedly will this code be used anywhere anymore?

**Review by BannAgainVA — COMMENTED** [review]
Needed
